library/swift/src/GRPCResponseHandler.swift

```diff
@@ -0,0 +1,158 @@
+import Foundation
+
+/// Handler for responses sent over gRPC.
+@objcMembers
+public final class GRPCResponseHandler: NSObject {
+  /// Represents the state of a response stream's body data.
+  private enum State {
+    /// Awaiting a gRPC compression flag.
+    case expectingCompressionFlag
+    /// Awaiting the length specification of the next message.
+    case expectingMessageLength
+    /// Awaiting a message with the specified length.
+    case expectingMessage(length: UInt32)
+  }
+
+  /// Underlying response handler which should be called with response data.
+  let underlyingHandler: ResponseHandler
+
+  /// Initialize a new instance of the handler.
+  ///
+  /// - parameter queue: Dispatch queue upon which callbacks will be called.
+  public init(queue: DispatchQueue = .main) {
+    self.underlyingHandler = ResponseHandler(queue: queue)
+  }
+
+  /// Specify a callback for when response headers are received by the stream.
+  /// If `endStream` is `true`, the stream is complete.
+  ///
+  /// - parameter closure: Closure which will receive the headers, gRPC status,
+  ///                      and flag indicating if the stream is headers-only.
+  @discardableResult
+  public func onHeaders(_ closure:
+    @escaping (_ headers: [String: [String]], _ grpcStatus: Int, _ endStream: Bool) -> Void)
```

Comment:
Since GRPC is guaranteed to send trailers with every response, you maybe don't need any `endStream` booleans on the callbacks.

Reply:
Nice. Based on gRPC [docs](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md):

> Response-Headers & Trailers-Only are each delivered in a single HTTP2 HEADERS frame block. Most responses are expected to have both headers and trailers but Trailers-Only is permitted for calls that produce an immediate error. Status must be sent in Trailers even if the status code is OK.

So we should be able to remove the `endStream` from `onMessage`